source/extensions/tracers/zipkin/span_context_extractor.cc

```diff
@@ -0,0 +1,227 @@
+bool validSamplingFlags(char c) {
+  if (c == '1' || c == '0' || c == 'd')
+    return true;
+  return false;
+}
+
+bool getSamplingFlags(char c, const Tracing::Decision tracing_decision) {
+  if (validSamplingFlags(c)) {
+    return c == '0' ? false : true;
+  } else {
+    return tracing_decision.traced;
+  }
+}
+
+} // namespace
+
+SpanContextExtractor::SpanContextExtractor(Http::HeaderMap& request_headers)
+    : request_headers_(request_headers) {}
+
+SpanContextExtractor::~SpanContextExtractor() {}
+
+bool SpanContextExtractor::extractSampled(const Tracing::Decision tracing_decision) {
+  bool sampled(false);
+  auto b3_header_entry = request_headers_.get(ZipkinCoreConstants::get().B3);
+  if (b3_header_entry) {
+    absl::string_view b3 = b3_header_entry->value().getStringView();
+    int sampled_pos = 0;
+    switch (b3.length()) {
+    case 1:
+      break;
+    case 35: // 16 + 1 + 16 + 2
+      sampled_pos = 34;
+      break;
+    case 51: // 32 + 1 + 16 + 2
+      sampled_pos = 50;
+      break;
+    case 52: // 16 + 1 + 16 + 2 + 1 + 16
+      sampled_pos = 34;
+      break;
+    case 68: // 32 + 1 + 16 + 2 + 1 + 16
+      sampled_pos = 50;
+      break;
+    default:
+      return tracing_decision.traced;
+    }
+    return getSamplingFlags(b3[sampled_pos], tracing_decision);
+  }
+
+  auto x_b3_sampled_entry = request_headers_.get(ZipkinCoreConstants::get().X_B3_SAMPLED);
+  if (!x_b3_sampled_entry) {
+    return tracing_decision.traced;
+  }
+  // Checking if sampled flag has been specified. Also checking for 'true' value, as some old
+  // zipkin tracers may still use that value, although should be 0 or 1.
+  absl::string_view xb3_sampled = x_b3_sampled_entry->value().getStringView();
+  sampled = xb3_sampled == ZipkinCoreConstants::get().SAMPLED || xb3_sampled == "true";
+  return sampled;
+}
+
+std::pair<SpanContext, bool> SpanContextExtractor::extractSpanContext(bool is_sampled) {
+  if (request_headers_.get(ZipkinCoreConstants::get().B3)) {
+    return extractSpanContextFromB3SingleFormat(is_sampled);
+  }
+  uint64_t trace_id(0);
+  uint64_t trace_id_high(0);
+  uint64_t span_id(0);
+  uint64_t parent_id(0);
+
+  auto b3_trace_id_entry = request_headers_.get(ZipkinCoreConstants::get().X_B3_TRACE_ID);
+  auto b3_span_id_entry = request_headers_.get(ZipkinCoreConstants::get().X_B3_SPAN_ID);
+  if (b3_span_id_entry && b3_trace_id_entry) {
+    // Extract trace id - which can either be 128 or 64 bit. For 128 bit,
+    // it needs to be divided into two 64 bit numbers (high and low).
+    const std::string tid = b3_trace_id_entry->value().c_str();
+    if (b3_trace_id_entry->value().size() == 32) {
+      const std::string high_tid = tid.substr(0, 16);
+      const std::string low_tid = tid.substr(16, 16);
+      if (!StringUtil::atoul(high_tid.c_str(), trace_id_high, 16) ||
+          !StringUtil::atoul(low_tid.c_str(), trace_id, 16)) {
+        throw ExtractorException(
```

Comment:
Please make sure this new file has 100% test coverage w/ all error cases tested. I think you are missing some.